src/common/transformations/src/transformations/common_optimizations/pull_through_reduce.cpp

```diff
@@ -0,0 +1,245 @@
+// Copyright (C) 2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "transformations/common_optimizations/pull_through_reduce.hpp"
+
+#include <memory>
+#include <vector>
+
+#include "itt.hpp"
+#include "openvino/core/validation_util.hpp"
+#include "openvino/op/util/reduction_base.hpp"
+#include "openvino/opsets/opset9.hpp"
+#include "openvino/pass/pattern/op/or.hpp"
+#include "openvino/pass/pattern/op/wrap_type.hpp"
+#include "sequnce_generator.hpp"
+#include "transformations/utils/utils.hpp"
+
+namespace {
+// Adjust axes of Unsqueeze/Reduce ops after Unsqueeze pulling
+// For example if we have:
+//          input(shape={5,10,15})
+//                 |
+//          Unsqueeze(axes=[0,3]) -> output_shape = {1,5,10,1,15}
+//                 |
+// ReduceOp(axes=[2,4], keep_dims=false) -> output_shape = {1,5,1}
+// after pulling it will be:
+//          input(shape={5,10,15})
+//                 |
+// ReduceOp(axes=[1,2], keep_dims=false) -> output_shape = {5}
+//                 |
+//          Unsqueeze(axes=[0,2]) -> output_shape = {1,5,1}
+const std::vector<int64_t> adjust_axes(const std::vector<int64_t>& axes_to_align,
+                                       const std::vector<int64_t>& offset_axes) {
+    auto number_of_axes_less_than = [&offset_axes](const int64_t current_axis) {
+        return std::count_if(std::begin(offset_axes),
+                             std::end(offset_axes),
+                             [&current_axis](const int64_t excluded_axis) {
+                                 return excluded_axis < current_axis;
+                             });
+    };
+    std::vector<int64_t> result;
+    for (const auto& axis : axes_to_align) {
+        result.push_back(axis - number_of_axes_less_than(axis));
+    }
+    return result;
+}
+
+// Try to represent given Reshape node via Unsqueeze and calculate axes of such Unsqueeze
+// The function works only for cases where the "1" dimensions are inserted on the left or right sides, like:
+// - Reshape(input_shape={5,10,15}, target_shape={5,10,15,1}), 3 axis returned
+// - Reshape(input_shape={5,10,15}, target_shape={1,5,10,15}), 0 axis returned
+// - Reshape(input_shape={5,10,15}, target_shape={1,5,10,15,1}), 0 and 3 axes returned
+// Cases where the axes are inserted in the middle, like:
+// - Reshape(input_shape={5,10,15}, target_shape={5,10,1,15})
+// are not supported now (ticket 94373)
+// For example if we have Reshape(input_shape={5,10,15}, target_shape={1,5,10,15,1}),
+// it can be represented via Unsqueeze with axes=[0,3]
+std::vector<int64_t> try_get_unsqueeze_axes_from_reshape(const ov::Shape& reshape_shape,
+                                                         const ov::Shape& reduce_shape) {
+    if (reshape_shape.size() < reduce_shape.size()) {
+        return {};
+    }
+    const auto it = std::search(std::begin(reshape_shape),
```

Comment:
Let's give it a chance to support cases in the middle: 

```
auto cur_element_idx = 0;
auto current_element = reduce_shape[cur_element_idx];
auto reshape_shape_idx = 0;
for (; reshape_shape_idx  < reshape_shape.size(); ++reshape_shape_idx) {
  if (current_element == reshape_shape[i] && current_element_idx + 1 < reduce_shape.size()) {
      ++cur_element_idx;
      current_element = reduce_shape[cur_element_idx];
  } else if (reshape_shape[i] == 1) {
     // it's ok
     continue;
  } else {
     break;
  }
}

if (current_element_idx == reduce_shape.size() -1 && reshape_shape_idx == reshape_shape.size()) {
   // we have found the pattern
} else {
   // our search has failed
}
```

Reply:
> Let's give it a chance to support cases in the middle:
> 
> ```
> auto cur_element_idx = 0;
> auto current_element = reduce_shape[cur_element_idx];
> auto reshape_shape_idx = 0;
> for (; reshape_shape_idx  < reshape_shape.size(); ++reshape_shape_idx) {
>   if (current_element == reshape_shape[reshape_shape_idx] && current_element_idx + 1 < reduce_shape.size()) {
>       ++cur_element_idx;
>       current_element = reduce_shape[cur_element_idx];
>   } else if (reshape_shape[reshape_shape_idx ] == 1) {
>      // it's ok
>      continue;
>   } else {
>      break;
>   }
> }
> 
> if (current_element_idx == reduce_shape.size() -1 && reshape_shape_idx == reshape_shape.size()) {
>    // we have found the pattern
> } else {
>    // our search has failed
> }
> ```

@itikhono Done, but modify to handle cases when `input_shape` and `target_shape` have "1"